Pythonevalobj

Necr0x0Der opened the pull request:
PythonEval fix to support grounding in object methods in accordance with #42

ngeiswei commented:
@Necr0x0Der you usually don't need to merge the master into your branch, unless you have a specific reason for it (such as you need a feature from the master to make sense of your work). No big deal of course, just saying for the next PRs.

I'm unqualified to review that PR, BTW (not strong on python bindings).

Necr0x0Der commented:
@ngeiswei , actually, I had to, because otherwise opencog's CicleCI didn't pass, because at the same time when I was raising a PR, you had pushed some changes, which affected unit tests, and these unit tests didn't pass for my PR. The simplest way for me was to update my branch in accordance to the upstream once again (although I did this exactly before rising PR).

vsbogd commented on opencog/cython/PythonEval.cc:
```diff
@@ -715,35 +716,71 @@ int PythonEval::argument_count(PyObject* pyFunction)
+        // Else find it as Attr in Object
+        pyObject = nullptr == pyObject ?
+            find_object(pyModule, objectName) :
+            PyObject_GetAttrString(pyObject, objectName.c_str());
```
```PyObject_GetAttrString()``` returns new reference, so you need to do ```Py_DECREF()``` for returned object after it is not needed. And ```find_object()``` returns result of ```PyDict_GetItemString()``` which is borrowed reference. So reference counter is different for first and second option.

vsbogd replied:
As an idea about managing memory here and below: you can create some small class which encapsulates all details which are required to call a function and use it instead of triple pyModule, pyObject, functionname. Then ```module_for_function``` will initialize this object and you can calculate and return number of arguments from it and so on.

Simple way is to wrap ```find_object``` by ```Py_INCREF``` or ```Py_XINCREF```

vsbogd commented on opencog/cython/PythonEval.cc:
```diff
@@ -776,14 +815,18 @@ PyObject* PythonEval::call_user_function(const std::string& moduleFunction,
-    PyObject* pyUserFunc = PyDict_GetItemString(pyDict, functionName.c_str());
+        pyUserFunc = PyDict_GetItemString(pyDict, functionName.c_str());
+    } else {
+        pyUserFunc = PyObject_GetAttrString(pyObject, functionName.c_str());
```
Same thing as in https://github.com/singnet/atomspace/pull/45/files#r235781526

vsbogd commented on opencog/cython/PythonEval.cc:
```diff
@@ -1027,7 +1072,12 @@ void PythonEval::apply_as(const std::string& moduleFunction,
+    if(nullptr == pyObject) {
+        pyUserFunc = PyDict_GetItemString(pyDict, functionName.c_str());
+    } else {
+        pyUserFunc = PyObject_GetAttrString(pyObject, functionName.c_str());
```
same as https://github.com/singnet/atomspace/pull/45/files#r235781526

noskill commented on opencog/cython/PythonEval.cc:
```diff
@@ -1055,6 +1105,8 @@ void PythonEval::apply_as(const std::string& moduleFunction,
     // Get the expected argument count.
     int expectedArgumentCount = this->argument_count(pyUserFunc);
+    // If there is an object, its method has 'self' argument, which is passed automatically
+    if (pyObject) expectedArgumentCount--;
```
Could fail on static methods, i guess.. Python3 has PyInstanceMethod_Check to check if it is instance method. On the other hand there is PyMethod_GET_SELF for python3 and  python2 as well which returns nullptr if method doesn't has self.

Necr0x0Der replied:
@noskill . Good point, but both these functions don't work. `PyInstanceMethod_Check` always returns null, while `PyMethod_GET_SELF` always returns something.

noskill replied:
Sorry for misleading you. I found that **PyMethod_Check** works, PyInstanceMethod_Check checks for something else and PyMethod_GET_SELF is made to return something anyway.

noskill replied:
I suggest to change this line to:
```c++
    if (pyObject and PyMethod_Check(pyUserFunc)){
            expectedArgumentCount--;
    }  
```

Necr0x0Der commented:
> May be I am missing something but `pyObject` will leak if exception is thrown above in line 855, 863 or 871. Looks like `pyUserFunc` has the same issue for lines 863, 871 and 884 at least.

It seems so. I moved `Py_DECREF(pyObject)` higher in the code and fixed the old issue with `pyUserFunc`, which was there for years.

vsbogd commented:
I have merged master as branch required update.
@ngeiswei what do you think, would it be better to rebase branch commits on master?

ngeiswei commented:
@vsbogd I'm fine either way. I tend to rebase when the feature branch is very long (like multiple months) just to avoid creating very long parallel branches in the git history but that's almost a matter of taste I guess.

ngeiswei commented:
I think the main situation where rebasing is undesirable is when you want to be sure that the state of the current code is preserved by a given commit on the feature branch cause you might need to go back to that exact same state. This can happen when you say keep a dairy, such as "in commit XXXXX we got these results", then rebasing may mess that up.

vsbogd commented:
Git suggested the button "Update branch" and I could not hold out of pressing it :-) But finally it just merged master to this branch.

vsbogd commented:
I will merge it then.

ngeiswei commented:
@vsbogd maybe it's because you've merged master into the branch... Anyway, our habit is to systematically merge via the github interface. But if one wishes to rebase onto the master prior to create a PR it's up to him/her.
